### PR TITLE
AccountsDb: make write_accounts_to_cache() never block sending to the bg hasher

### PR DESCRIPTION
**Pull request by alessandrod**
Avoid commit_transactions() => store() => store_accounts_to() callers having to notify the bg hasher thread by doing an explicit sleep instead of sleeping on channel.recv() when the channel is empty. This avoids the case in which multiple replay/banking threads call commit_transactions() at the same time and end up... sleeping themselves acquiring the mutex to wake up the bg hasher.

Right side of the marker is with this PR

Screenshot 2024-11-28 at 9 45 32 am

### CODE REVIEW

**Comment by alessandrod** [comment]
Here's a run I did a few weeks ago at higher throughput 

Screenshot 2024-12-02 at 10 21 36 pm

**Comment by brooksprumo** [comment]
Can you help me understand, please?

On the `store` side, here's where we send the accounts to the background thread for hashing:
https://github.com/anza-xyz/agave/blob/c11fe18aff967b86705f3eb421f96215671d0bae/accounts-db/src/accounts_db.rs#L6531-L6534

Are you saying that `send()` blocks these threads because we are calling `recv()` instead of `try_recv()` in the background hasher thread here?

https://github.com/anza-xyz/agave/blob/c11fe18aff967b86705f3eb421f96215671d0bae/accounts-db/src/accounts_db.rs#L2347

**Comment by brooksprumo** [comment]
Or is the issue because `send()` will notify receivers, and using `recv()` is what registers a SyncReceiver?

notify receivers in send():
https://docs.rs/crossbeam-channel/latest/src/crossbeam_channel/flavors/list.rs.html#295-296

grab a lock to wake up a receiver:
https://docs.rs/crossbeam-channel/latest/src/crossbeam_channel/waker.rs.html#223

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -2355,7 +2355,10 @@ impl AccountsDb {
                         };
                     }
                 }
-                Err(err) => {
+                Err(TryRecvError::Empty) => {
+                    sleep(Duration::from_millis(5));
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:2359]: I'm not sure what the "right" sleep duration should be here, but 5 ms seems fine to start with.

**Review by brooksprumo — APPROVED** [review]
:shipit: 

(assuming my understanding in https://github.com/anza-xyz/agave/pull/3814#issuecomment-2511664549 is correct)

**Comment by alessandrod** [comment]
> Or is the issue because `send()` will notify receivers, and using `recv()` is what registers a SyncReceiver?
> 
> notify receivers in send(): https://docs.rs/crossbeam-channel/latest/src/crossbeam_channel/flavors/list.rs.html#295-296
> 
> grab a lock to wake up a receiver: https://docs.rs/crossbeam-channel/latest/src/crossbeam_channel/waker.rs.html#223

correct, the issue is racing to notify, sleeping on the lock

**Pull request merged**